crates/sdl-encoder/src/directive_def.rs

```diff
@@ -0,0 +1,170 @@
+use crate::InputValue;
+use std::fmt::{self, Display};
+
+/// The `__Directive` type represents a Directive that a service supports.
+///
+/// *DirectiveDefinition*:
+///     Description<sub>opt</sub> **directive @** Name Arguments Definition<sub>opt</sub> **repeatable**<sub>opt</sub> **on** DirectiveLocations
+///
+/// Detailed documentation can be found in [GraphQL spec](https://spec.graphql.org/draft/#sec-Type-System.Directives).
+///
+/// ### Example
+/// ```rust
+/// use sdl_encoder::{Directive};
+/// use indoc::indoc;
+///
+/// let mut directive = Directive::new("infer".to_string());
+/// directive.description(Some("Infer field types\nfrom field values.".to_string()));
+/// directive.location("OBJECT".to_string());
+/// directive.location("FIELD_DEFINITION".to_string());
+/// directive.location("INPUT_FIELD_DEFINITION".to_string());
+///
+/// assert_eq!(
+///     directive.to_string(),
+///     r#""""
+/// Infer field types
+/// from field values.
+/// """
+/// directive @infer on OBJECT | FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+/// "#
+/// );
+/// ```
+#[derive(Debug)]
+pub struct Directive {
+    // Name must return a String.
+    name: String,
+    // Description may return a String or null.
+    description: Option<String>,
+    // Args returns a Vector of __InputValue representing the arguments this
+    // directive accepts.
+    args: Vec<InputValue>,
+    // Locations returns a List of __DirectiveLocation representing the valid
+    // locations this directive may be placed.
+    locations: Vec<String>,
+}
+
+impl Directive {
+    /// Create a new instance of Directive definition.
+    pub fn new(name: String) -> Self {
+        Self {
+            name,
+            description: None,
+            args: Vec::new(),
+            locations: Vec::new(),
+        }
+    }
+
+    /// Set the Directive's description.
+    pub fn description(&mut self, description: Option<String>) {
+        self.description = description;
+    }
+
+    /// Set the Directive's location.
+    pub fn location(&mut self, location: String) {
+        self.locations.push(location);
+    }
+
+    /// Set the Directive's args.
+    pub fn arg(&mut self, arg: InputValue) {
+        self.args.push(arg);
+    }
+}
+
+impl Display for Directive {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            // We are determing on whether to have description formatted as
+            // a multiline comment based on whether or not it already includes a
+            // \n.
+            match description.contains('\n') {
+                true => writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?,
+                false => writeln!(f, "\"\"\"{}\"\"\"", description)?,
+            }
```

Comment:
Maybe extract the multiline formatting code into a separate function to avoid code duplication?